docs/conceptual/core/codec.md

```diff
@@ -1,13 +1,181 @@
+
+#### Example
+
+- `signed 8-bit integer 69`: `0x45`
+- `unsigned 16-bit integer 42`: `0x2a00`
+- `unsigned 32-bit integer 16777215`: `0xffffff00`
+
+### Compact/General Integers
+
+A "compact" or general integer encoding is sufficient for encoding large
+integers (up to 2**536) and is more efficient at encoding most values than the
+fixed-width version. (Though for single-byte values, the fixed-width integer is
+never worse.)
+
+It is encoded with the two least significant bits denoting the mode:
+
+- `0b00`: single-byte mode; upper six bits are the LE encoding of the value
+  (valid only for values of 0-63).
+- `0b01`: two-byte mode: upper six bits and the following byte is the LE
+  encoding of the value (valid only for values `64-(2**14-1)`).
+- `0b10`: four-byte mode: upper six bits and the following three bytes are the
+  LE encoding of the value (valid only for values `(2**14-1)-(2**30-1)`).
+- `0b11`: Big-integer mode: The upper six bits are the number of bytes
+  following, less four. The value is contained, LE encoded, in the bytes
+  following. The final (most significant) byte must be non-zero. Valid only for
+  values `(2**30-1)-(2**536-1)`.
+
+#### Example
+
+- `unsigned integer 0`: `0x00`
+- `unsigned integer 1`: `0x04`
+- `unsigned integer 42`: `0xa8`
+- `unsigned integer 69`: `0x1501`
+
+Error:
+
+- ~~`0x0100`: Zero encoded in mode 1~~
+
+### Boolean
+
+Boolean values are encoded using the least significant bit (LSB) of a single
+byte.
+
+#### Example
+
+- `boolean false`: `0x00`
+- `boolean true`: `0x01`
+
+### Options
+
+One or zero values of a particular type. Encoded as:
+
+- `0x00` if it is `None` ("empty" or "null").
+- `0x01` followed by the encoded value if it is `Some`.
+
+As an exception, in the case that the type is a boolean, then it is always one
+byte:
+
+- `0x00` if it is `None` ("empty" or "null").
+- `0x01` if it is the `false` value.
+- `0x02` if it is the `true` value.
+
+### Vectors (lists, series, sets)
+
+A collection of same-typed values is encoded, prefixed with a *compact* encoding
+of the number of items, followed by each item's encoding concatenated in turn.
+
+#### Example
+
+Vector of unsigned 16-bit integers:
+
+```
+[4, 8, 15, 16, 23, 42]
+```
+
+SCALE Bytes:
+
+```
+0x18040008000f00100017002a00
+```
+
+### Tuples and Structures
+
+A fixed-size series of values, each with a possibly different but predetermined
+and fixed type. This is simply the concatenation of each encoded value. For
+structures, the values are named, but that is irrelevant for the encoding (names
+are ignored - only order matters).
+
+#### Example
+
+Tuple of compact unsigned integer and boolean:
+
+`(3, false)`: `0x0c00`
+
+### Enumerations (tagged-unions)
+
+A fixed number of variants, each mutually exclusive and potentially implying a
+further value or series of values.
+
+Encoded as the first byte identifying the index of the variant that the value
+is. Any further bytes are used to encode any data that the variant implies.
```

Comment:
would it worth saying: thus no more than 256 variants are supported

Reply:
updated